### PR TITLE
JS: track document and DOM values with type tracking

### PR DESCRIPTION
**Pull request by asger-semmle**
This PR does a few things:
1. Use TypeTracking to track `document` and DOM values.
2. Move some predicates into the `DOM::` namespace in order to safely introduce the `documentRef()`, `locationRef()`, etc outside the global scope.
3. Rewrite some other DOM code to use `DataFlow` library and some of the new predicates that use type tracking.

[Evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/balbinus.ti.semmle.com_1553834869797) of the first commit alone shows some decent results, though the performance looks too good to be true.

I'll re-run a full evaluation over the weekend with latest changes, ideally including those from review.

There is still more modernization to be done in the DOM libraries, but I'd prefer not to do all of it in this PR.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/DOM.qll [hunk 1]
```diff
@@ -253,4 +253,86 @@ module DOM {
...
+      methodName = "createRange" or
+      methodName = "getElementById" or
+      methodName = "querySelector"
+    |
+      result = documentRef().getAMethodCall(methodName) or
+      result = DataFlow::globalVarRef(methodName).getACall()
+    )
+  }
+
+  /** Gets a data flow node that refer directly to a value from the DOM. */
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/DOM.qll]:
```suggestion
  /** Gets a data flow node that refers directly to a value from the DOM. */
```

File: javascript/ql/src/semmle/javascript/DOM.qll [hunk 2]
```diff
@@ -253,4 +253,86 @@ module DOM {
...
+  DataFlow::SourceNode domValueSource() {
+    result.asExpr().(VarAccess).getVariable() instanceof DOMGlobalVariable or
+    result = domValueRef().getAPropertyRead(_) or
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/DOM.qll]:
```suggestion
    result = domValueRef().getAPropertyRead() or
```

File: javascript/ql/src/semmle/javascript/DOM.qll [hunk 3]
```diff
@@ -253,4 +253,86 @@ module DOM {
...
+    t.start() and
+    result = domValueSource()
+    or
+    exists(DataFlow::TypeTracker t2 | result = domValueRef(t2).track(t2, t))
+  }
+
+  /** Gets a data flow node that may refer to a value from the DOM. */
+  DataFlow::SourceNode domValueRef() { result = domValueRef(DataFlow::TypeTracker::end()) }
+
+  /** Gets a data flow node that directly refers to the DOM `location` object. */
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/DOM.qll]:
```suggestion
  /** Gets a data flow node that directly refers to a DOM `location` object. */
```

Similarly elsewhere.

File: javascript/ql/src/semmle/javascript/DOM.qll [hunk 4]
```diff
@@ -253,4 +253,86 @@ module DOM {
...
+    result = locationSource()
+    or
+    exists(DataFlow::TypeTracker t2 | result = locationRef(t2).track(t2, t))
+  }
+
+  /** Gets a reference to the DOM `location` object. */
+  DataFlow::SourceNode locationRef() { result = locationRef(DataFlow::TypeTracker::end()) }
+
+  /**
+   * Gets a reference to the 'document' object.
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/DOM.qll]:
```suggestion
   * Gets a reference to the `document` object.
```

(and below)

**Comment by asger-semmle** [comment]
Thanks! All done.

**Comment by asger-semmle** [comment]
The [evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/augustus.ti.semmle.com_1554012514857) showed some modest slow down on 4 projects.

[Re-run](https://git.semmle.com/asger/dist-compare-reports/tree/augustus.ti.semmle.com_1554135487545) of the 10 slowest slugs.

I've tried caching some predicates and re-run it twice on the 10 slowest slugs ([first](https://git.semmle.com/asger/dist-compare-reports/tree/augustus.ti.semmle.com_1554121735016), [second](https://git.semmle.com/asger/dist-compare-reports/tree/augustus.ti.semmle.com_1554127131297)) which was overall a bit inconclusive.

It might just be the price we have to pay for having more sources and sinks.

**Comment by xiemaisi** [comment]
The slowdown seems like a price worth paying for the extra results we get.

**Pull request merged**